inference-engine/src/mkldnn_plugin/nodes/mkldnn_depth_to_space_node.cpp

```diff
@@ -0,0 +1,353 @@
+
+        size_t orderShiftForBlocks, orderShiftForDims;
+        if (mode == Mode::BLOCKS_FIRST) {
+            orderShiftForBlocks = 1;
+            orderShiftForDims = nSpatialDims + 2;
+
+            optimizedParams.src_block_dims[nSpatialDims + 1] = srcBlockedDims[1] / blockStep;
+            optimizedParams.src_block_dims[lastIdx] = srcBlockedDims.back();
+
+            order[1] = nSpatialDims + 1;
+            order[lastIdx] = lastIdx;
+        } else {
+            orderShiftForBlocks = nSpatialDims + 4;
+            orderShiftForDims = 3;
+
+            size_t newBlockSize = srcBlockedDims.back() / blockStep;
+            size_t newBlocksCount = srcBlockedDims[1] * newBlockSize / srcBlockedDims.back();
+            optimizedParams.src_block_dims[1] = newBlocksCount;
+            optimizedParams.src_block_dims[2] = srcBlockedDims[1] / newBlocksCount;
+            optimizedParams.src_block_dims[lastIdx - nSpatialDims] = newBlockSize;
+
+            order[1] = 1;
+            order[2] = 3;
+            order[lastIdx - 1] = 2;
+            order[lastIdx] = lastIdx - nSpatialDims;
+        }
+
+        reshapeAndSetPermOrder(orderShiftForDims, orderShiftForBlocks, firstSpatialOrder, srcBlockedDims);
+    } else if (getParentEdgeAt(0)->getMemory().GetDesc().isTailCFormat()) {
+        srcDims.push_back(srcDims[1]);
+        dstDims.push_back(dstDims[1]);
+        srcDims.erase(srcDims.begin() + 1);
+        dstDims.erase(dstDims.begin() + 1);
+        firstSpatialOrder = 1;
+
+        size_t shift = static_cast<size_t>(mode == DEPTH_FIRST) + nSpatialDims + 1;
+        order[lastIdx] = mode == Mode::DEPTH_FIRST ? nSpatialDims + 1 : lastIdx;
+        optimizedParams.src_block_dims[order[lastIdx]] = srcDims.back() / blockStep;
+
+        reshapeAndSetPermOrder(firstSpatialOrder, shift, firstSpatialOrder, srcDims);
+    } else {
+        size_t shift = static_cast<size_t>(mode == DEPTH_FIRST) + 1;
+        order[1] = mode == DEPTH_FIRST ? 1 : nSpatialDims + 1;
+        optimizedParams.src_block_dims[order[1]] = srcDims[1] / blockStep;
+
+        reshapeAndSetPermOrder(nSpatialDims + firstSpatialOrder, shift, firstSpatialOrder, srcDims);
+    }
+
+    prepareOptimizedParams(getSelectedPrimitiveDescriptor()->getConfig().inConfs[0].desc.getPrecision());
+    prepareConfigParams();
+}
+
+void MKLDNNDepthToSpaceNode::execute(mkldnn::stream strm) {
+    auto srcData = reinterpret_cast<const uint8_t*>(this->getParentEdgeAt(0)->getMemoryPtr()->GetPtr());
+    auto dstData = reinterpret_cast<uint8_t*>(this->getChildEdgeAt(0)->getMemoryPtr()->GetPtr());
+
+    int MB = batchToProcess();
+    if (params.shape5D[0] != MB)
+        params.shape5D[0] = MB;
+
+    if (permute_kernel) {
+        auto &jcp = (*permute_kernel).jcp;
+        if (jcp.dst_block_dims[0] != MB)
+            jcp.dst_block_dims[0] = MB;
+
+        optimizedExecute(srcData, dstData);
+        return;
+    }
+
+    if (getParentEdgeAt(0)->getMemory().GetDesc().isBlockedCFormat()) {
```

Comment:
In what cases permute_kernel cannot be created? We need to have reference implementation inside permute itself to be able to reuse it got ShaffleChannels for examples. Code blocks below should be deleted at all.

Reply:
Added Permute reference and used it for DepthToSpace and SpaceToDepth references for cases without permute_kernel. Thank you